File: swig/R/test_linalg.R
test_sort <- function() {
   x = rnorm(2000000,0,1)
   return(Xtest(quote(sort(x)),quote(spams.sort(x,TRUE))))
}

test_calcAAt <- function () {
  m = 200;n = 200000; d= 0.05
  A = rSpMatrix(m,n,floor(m * n * d))
  return(Xtest(quote(A %*% t(A)),quote(spams.calcAAt(A))))
}

test_calcXAt <- function () {
  m = 200;n = 200000; d= 0.05
  A = rSpMatrix(m,n,floor(m * n * d))
  X = matrix(rnorm(64 * n),nrow = 64,ncol = n)
  return(Xtest(quote(X %*% t(A)),quote(spams.calcXAt(X,A))))
}
matprod <- function(x,y) "%*%"(x,y)
test_calcXY <- function() {
    X = matrix(rnorm(64 * 200),nrow = 64,ncol = 200,byrow = FALSE)
    Y = matrix(rnorm(200 * 20000),nrow = 200,ncol = 20000,byrow = FALSE)
    return(Xtest(quote(X %*% Y),quote(spams.calcXY(X,Y))))
}

test_calcXYt <- function () {
    X = matrix(rnorm(64 * 200),nrow = 64,ncol = 200,byrow = FALSE)
    Y = matrix(rnorm(200 * 20000),nrow = 20000,ncol = 200,byrow = FALSE)
    return(Xtest(quote(X %*% t(Y)),quote(spams.calcXYt(X,Y))))
  
}

test_calcXtY <- function () {
    X = matrix(rnorm(64 * 200),nrow = 200,ncol = 64,byrow = FALSE)
    Y = matrix(rnorm(200 * 20000),nrow = 200,ncol = 20000,byrow = FALSE)
    return(Xtest(quote(t(X) %*% Y),quote(spams.calcXtY(X,Y))))
}

test_bayer <- function () {
  X = rnorm(2000000,0,1)
  y2 = Xtest1('spams',quote(spams.bayer(X,0)),n=1)
  return(NULL)
}

test_conjGrad <- function () {
  A = matrix(rnorm(500 * 5000),nrow = 5000,ncol = 500)
  A = t(A) %*% A
  b = as.vector(rep(1,ncol(A)))
  x0 = b
  tol = 1e-4
  itermax = floor(0.5 * length(b))
  CGtest <- function(txt,expr) {
    tic = proc.time()
    for (i in 1:20)
      y = eval(expr)
    tac = proc.time()
    cat(sprintf("  Time (%s): %f\n",txt,(tac - tic)[['user.self']]))
    x1 = abs(b - (A %*% y))
    cat(sprintf("Mean error on b : %f\n\n",sum(x1) / length(b)))
    return(y)
  }
  y2 = CGtest("R",quote(solve(A,b)))
  y1 = CGtest("spams",quote(spams.conjGrad(A,b,x0,tol,itermax)))
  return(max(abs(y1 - y2)))
}

test_invSym <- function() {
    A = matrix(runif(1000 * 1000,0,1),nrow = 1000,ncol = 1000,byrow = FALSE)
    A = t(A) %*% A
    return(Xtest(quote(solve(A)),quote(spams.invSym(A))))
}
test_normalize <- function() {
    A = matrix(runif(100 * 1000,0,1),nrow = 100,ncol = 1000)
    y2 = Xtest1("spams",quote(spams.normalize(A)),n=1)
    return(NULL)
}

test_linalg.tests =list( 'sort' = test_sort,
  'calcAAt' = test_calcAAt,
  'calcXAt' = test_calcXAt,
  'calcXY' = test_calcXY,
  'calcXYt' = test_calcXYt,
  'calcXtY' = test_calcXtY,
  'bayer' = test_bayer,
  'conjGrad' = test_conjGrad,
  'invSym' = test_invSym,
  'normalize' = test_normalize
  )

File: test/test_linalg.py
import numpy as np
import scipy.sparse as ssp
import scipy.linalg
import spams
import time

ssprand = ssp.rand

import pytest
from numpy.testing import assert_allclose

@pytest.mark.parametrize("myfloat", [np.float32, np.float64])
def test_sort(myfloat):
    n = 2000000
    X = np.asfortranarray(np.random.normal(size=(n,)), dtype=myfloat)
    return assert_allclose(np.sort(X), spams.sort(X,True))


@pytest.mark.parametrize("myfloat", [np.float64])
def test_calcAAt(myfloat):
    """
    test A * A'
    """
    m = 200
    n = 200000
    d = 0.05
    A = ssprand(m, n, density=d, format="csc", dtype=myfloat)
    return assert_allclose((A @ A.T).todense(), spams.calcAAt(A))


@pytest.mark.parametrize("myfloat", [np.float64])
def test_calcXAt(myfloat):
    m = 200
    n = 200000
    d = 0.05
    A = ssprand(m, n, density=d, format="csc", dtype=myfloat)
    X = np.asfortranarray(np.random.normal(size=(64, n)), dtype=myfloat)
    return assert_allclose((X @ A.T) , spams.calcXAt(X,A))


@pytest.mark.parametrize("myfloat", [np.float64])
def test_calcXY(myfloat):
    X = np.asfortranarray(np.random.normal(size=(64, 200)), dtype=myfloat)
    Y = np.asfortranarray(np.random.normal(size=(200, 20000)), dtype=myfloat)
    return assert_allclose(np.dot(X,Y), spams.calcXY(X,Y))


@pytest.mark.parametrize("myfloat", [np.float64])
def test_calcXYt(myfloat):
    X = np.asfortranarray(np.random.normal(size=(64, 200)), dtype=myfloat)
    Y = np.asfortranarray(np.random.normal(size=(20000, 200)), dtype=myfloat)
    return assert_allclose(np.dot(X,Y.T), spams.calcXYt(X,Y))


@pytest.mark.parametrize("myfloat", [np.float64])
def test_calcXtY(myfloat):
    X = np.asfortranarray(np.random.normal(size=(200, 64)), dtype=myfloat)
    Y = np.asfortranarray(np.random.normal(size=(200, 20000)), dtype=myfloat)
    return assert_allclose(np.dot(X.T,Y), spams.calcXtY(X,Y))


@pytest.mark.parametrize("myfloat", [np.float32, np.float64])
def test_bayer(myfloat):
    n = 2000000
    X = np.asfortranarray(np.random.normal(size=(n,)), dtype=myfloat)

    Z = spams.bayer(X,0)
    return None


@pytest.mark.parametrize("myfloat", [np.float32, np.float64])
def test_conjGrad(myfloat):
    A = np.asfortranarray(np.random.normal(size=(5000, 500)))
    # *    np.random.seed(0)
    # *    A = np.asfortranarray(np.random.normal(size = (10,5)))
    A = np.asfortranarray(np.dot(A.T, A), dtype=myfloat)
    b = np.ones((A.shape[1],), dtype=myfloat, order="F")
    x0 = b
    tol = 1e-4
    itermax = int(0.5 * len(b))

    tic = time.time()
    for i in range(0, 20):
        y1 = np.linalg.solve(A, b)
    tac = time.time()
    print("  Time (numpy): ", tac - tic)
    x1 = np.abs(b - np.dot(A, y1))
    print("Mean error on b : %f" % (x1.sum() / b.shape[0]))

    tic = time.time()
    for i in range(0, 20):
        y2 = spams.conjGrad(A, b, x0, tol, itermax)
    # *        y2 = spams.conjGrad(A,b)
    tac = time.time()
    print("  Time (spams): ", tac - tic)
    x1 = np.dot(A, y2)
    x2 = np.abs(b - x1)
    print("Mean error on b : %f" % (x2.sum() / b.shape[0]))

    err = np.abs(y1 - y2)
    print("Max error", err.max())


@pytest.mark.parametrize("myfloat", [np.float64])
def test_invSym(myfloat):
    A = np.asfortranarray(np.random.random(size=(1000, 1000)))
    A = np.asfortranarray(np.dot(A.T, A), dtype=myfloat)
    assert_allclose(spams.invSym(A), scipy.linalg.pinvh(A), atol=1e-5)


@pytest.mark.parametrize("myfloat", [np.float32, np.float64])
def test_normalize(myfloat):
    A = np.asfortranarray(np.random.random(size=(100, 1000)), dtype=myfloat)
    res2 = spams.normalize(A)
    return None
